### PR TITLE
fix(tournaments): persist game/regulation settings + 7 related bugs

### PR DESCRIPTION
**Pull request by thatguyinabeanie**
## Summary

The tournament settings page (`/dashboard/community/[slug]/tournaments/[slug]/manage/settings`) silently dropped game / regulation / platform / battle_format on every refresh because the page client built `tournamentForSettings` without those fields — the form fell back to its hardcoded defaults (`"sv"` / `"reg-i"` / `"cartridge"` / `"doubles"`) on every load. Fixed alongside seven related issues found while tracing the save flow.

## Fixes

| # | Issue | Fix |
| --- | --- | --- |
| 1 | Game/regulation/platform/battle_format dropped from page-client props | Added the four fields to the props object passed to `<TournamentSettings />` |
| 2 | Saves bypassed cache invalidation — public tournament page stayed stale | Routed updates through the existing `updateTournament` server action (widened to accept the full settings payload). Same for delete — caches now invalidate after a successful delete (so a thrown mutation doesn't evict tags) |
| 3 | Cancel after a save reverted to pre-save values | Added `savedFormData` + `savedPhaseConfigs` snapshots that advance on every successful save. `useSupabaseQuery` doesn't refetch on `router.refresh()`, so the prop alone is unreliable |
| 4 | `endDate` (and `startDate`) couldn't be cleared once set | Form sends `null` on clear; mutation/action types widened to `string \| null` |
| 5 | Silent overwrite of null game settings with `"sv"`/`"reg-i"` defaults | `buildUpdatePayload` diffs against the saved snapshot — only changed fields are sent. Saving with no edits is now a no-op |
| 6 | Native `window.confirm()` for delete | Replaced with the project's `AlertDialog` (matches `confirm-disconnect-dialog`) |
| 7 | Phases double-fetched | Dropped the separate `getTournamentPhases` query — `getTournamentBySlug` already returns `phases` joined |
| 8 | Delete redirected to public `/tournaments` | Now redirects to `/dashboard/community/[slug]/tournaments` |

## Files

- `apps/web/src/app/(dashboard)/.../manage/settings/tournament-settings-page-client.tsx` — pass the four missing fields, drop the duplicate phases query, pass `communitySlug` for the post-delete redirect
- `apps/web/src/components/tournaments/manage/tournament-settings.tsx` — Server-Action save/delete, saved-snapshot diffing, AlertDialog delete UX, null-aware date handling
- `apps/web/src/actions/tournaments.ts` — widen `updateTournament` payload; both `updateTournament` and `deleteTournament` now invalidate tournament + list + community caches
- `packages/supabase/src/mutations/tournaments/tournament-crud.ts` — accept `string | null` for dates
- Tests updated to mock the new server-action call sites and the AlertDialog delete flow

## Test plan

- [x] `pnpm typecheck` — clean across all 11 packages
- [x] `pnpm --filter @trainers/web test` — 4568 tests pass (1 new test for the no-diff skip; 6 updated for AlertDialog + Server Action flow + broadened invalidation)
- [x] `pnpm --filter @trainers/supabase test` — 1211 tests pass
- [x] `pnpm --filter @trainers/web lint` and `@trainers/supabase lint` — clean
- [ ] CI to run E2E
- [ ] Manual smoke: edit game + regulation, save, refresh — values persist
- [ ] Manual smoke: edit + save, click Edit again, change something else, click Cancel — first save's values are still there
- [ ] Manual smoke: clear end date — saves as null
- [ ] Manual smoke: delete confirmation uses AlertDialog and redirects to community dashboard

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Expanded tournament settings: game format, platform, battle format, registration/check-in options, nullable dates/participants, and links to manage pages from Upcoming Tournaments.
  * Modal confirmation for deletion and redirect to community tournaments after delete.

* **Bug Fixes**
  * Save now sends only changed fields and preserves saved snapshots; Cancel reliably restores last saved state.
  * Stronger validation prevents invalid updates and shows clear validation errors.
  * Improved cache revalidation after updates/deletes.

* **Tests**
  * Extensive new/updated tests covering validation, cache invalidation, save/delete flows, auth edge cases, and navigation.

### CODE REVIEW

**Comment by thatguyinabeanie** [comment]
Addressed all three inline review findings in 9fb011b9:

**Copilot — `handleSave` race condition (line 311):**
Snapshot live `formData` / `phaseConfigs` into local consts at save start. The action call, validation, and snapshot advance now all use the captured values, so a mid-save keystroke can't desync the saved snapshot from what was persisted.

**CodeRabbit — `playerCapEnabled` treats undefined as enabled (line 139):**
`!== null` → `!= null`. An omitted `max_participants` prop is now treated as uncapped, matching the documented semantics.

**CodeRabbit (outside-diff) — Zod validation at the action boundary:**
Extended `updateTournamentSchema` in `@trainers/validators` to cover every settings field (game / format / platform / battleFormat / dates / cap / registration / late-round). The `updateTournament` server action now `safeParse`s before calling the mutation; validation failures return `ActionResult { success: false, validationErrors }` without touching the database. New tests pin down: short-name rejection, out-of-range cap rejection, and full-payload acceptance.

The other CodeRabbit outside-diff finding (delete cache race) was already addressed in `bf86946c` — the action now captures community context first, deletes, then invalidates only on success.

File: apps/web/src/components/tournaments/manage/tournament-settings.tsx [hunk 1]
```diff
@@ -215,6 +264,10 @@ export function TournamentSettings({
         description: "Tournament settings have been updated successfully.",
       });
 
+      // Advance the saved snapshots so a follow-up edit + cancel reverts to
+      // the just-saved values, not the stale prop.
+      setSavedFormData(formData);
+      setSavedPhaseConfigs(phaseConfigs);
       setIsEditing(false);
       router.refresh();
```
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/components/tournaments/manage/tournament-settings.tsx:345]: Fixed in commit 9fb011b9 — `handleSave` now snapshots `formData` and `phaseConfigs` into local consts at save start. The action call, validation, snapshot advance, and phases save all use the captured values, so a mid-save keystroke can't desync the saved snapshot from what was actually persisted.

File: apps/web/src/components/tournaments/manage/tournament-settings.tsx [hunk 2]
```diff
@@ -104,52 +130,44 @@ export function TournamentSettings({
     battleFormat: (tournament.battle_format as BattleFormat) || "doubles",
     maxParticipants: tournament.max_participants?.toString() || "",
     playerCapEnabled: tournament.max_participants !== null,
```
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/components/tournaments/manage/tournament-settings.tsx]: Fixed in commit 9fb011b9 — `!== null` → `!= null`. An omitted `max_participants` prop is now treated as uncapped (matches the documented `number | null | undefined` semantics).

File: packages/validators/src/tournament.ts [hunk 3]
```diff
@@ -48,12 +48,56 @@ export const createTournamentSchema = z.object({
...
+  startDate: isoDateTimeOrNullSchema.optional(),
+  endDate: isoDateTimeOrNullSchema.optional(),
+  // 4..512 mirrors the input min/max in the settings UI.
+  maxParticipants: z.number().int().min(4).max(512).nullable().optional(),
+  status: tournamentStatusSchema.optional(),
+  // Game settings — kept as strings (DB columns are text); concrete values are
+  // managed by `@trainers/pokemon` and may evolve as new games/regs ship.
+  game: z.string().min(1).max(50).optional(),
+  gameFormat: z.string().min(1).max(50).optional(),
+  platform: z.enum(["cartridge", "showdown"]).optional(),
```
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply packages/validators/src/tournament.ts]: Fixed in commit 91efef59 — extracted an `optionalTrimmedString(max)` helper that follows the established `optionalHandle` / `optionalUrl` pattern in `organization-request.ts`: `.transform(v => v.trim())` → `.pipe(z.string().min(1).max(N))` → empty-or-undefined. Applied to `format`, `game`, and `gameFormat`. Added it.each tests for whitespace-only rejection and surrounding-whitespace trim.

[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply packages/validators/src/tournament.ts]: Fixed in same commit (91efef59) — see the reply on the CodeRabbit thread above.

File: apps/web/src/components/tournaments/manage/tournament-settings.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/components/tournaments/manage/tournament-settings.tsx]: Thanks for the ack! 🙏

File: apps/web/src/actions/tournaments.ts [hunk 4]
```diff
@@ -175,27 +177,30 @@ export async function createTournament(data: {
...
+    // bypass UI constraints (out-of-range cap, unknown registration type,
+    // malformed dates), and the underlying mutation does not reject them.
+    const parsed = updateTournamentSchema.safeParse(updates);
+    if (!parsed.success) {
+      return {
+        success: false,
+        error: "Invalid tournament settings",
+        validationErrors: parsed.error.issues.map((i) => i.message),
+      };
     }
```
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/actions/tournaments.ts:199]: Fixed in commit eb6f28d6 — added an explicit guard `if (Object.keys(parsed.data).length === 0) return { success: true }` after schema validation. The settings UI's `buildUpdatePayload` already filters unchanged fields, so this is defense-in-depth at the action boundary. Added a test that asserts an empty payload skips both the mutation and cache invalidation.

File: apps/web/src/components/tournaments/manage/tournament-settings.tsx [hunk 5]
```diff
@@ -169,36 +195,110 @@ export function TournamentSettings({
...
+    if (!data.name.trim()) return "Tournament name is required.";
+    if (data.playerCapEnabled) {
+      const cap = parseInt(data.maxParticipants, 10);
+      if (
+        Number.isNaN(cap) ||
+        cap < PLAYER_CAP_MIN ||
+        cap > PLAYER_CAP_MAX
+      ) {
+        return `Player cap must be between ${PLAYER_CAP_MIN} and ${PLAYER_CAP_MAX}.`;
+      }
```
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/components/tournaments/manage/tournament-settings.tsx:229]: Fixed in commit 73edb898 — replaced `parseInt(s, 10)` with a `parseStrictInt(value)` helper that uses `Number(s)` + `Number.isInteger()`. Decimals like `32.5` now fail validation with "must be a whole number" instead of silently saving `32`. Added `step={1}` on the number input. Test: "rejects a decimal player cap instead of silently truncating it".

File: apps/web/src/components/tournaments/manage/tournament-settings.tsx [hunk 6]
```diff
@@ -623,17 +716,17 @@ export function TournamentSettings({
...
+                    const parsed = parseInt(e.target.value, 10);
                     setFormData((prev) => ({
                       ...prev,
-                      lateCheckInMaxRound:
-                        parseInt(e.target.value) || undefined,
-                    }))
-                  }
+                      lateCheckInMaxRound: Number.isNaN(parsed) ? null : parsed,
+                    }));
+                  }}
```
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/components/tournaments/manage/tournament-settings.tsx:756]: Fixed in commit 73edb898 — same `parseStrictInt` helper applied to the late-round `onChange`. A pasted `3.7` clears the field instead of landing in state as `3`. Added `step={1}` on the input.

File: apps/web/src/actions/tournaments.ts [hunk 7]
```diff
@@ -175,27 +177,38 @@ export async function createTournament(data: {
...
+    // Short-circuit empty updates so we don't run a no-op DB write and bust
+    // every cache tag for nothing. The settings UI's `buildUpdatePayload`
+    // already filters unchanged fields, but a crafted request could still
+    // send `{}`.
+    if (Object.keys(parsed.data).length === 0) {
+      return { success: true, data: { success: true } };
     }
 
+    const supabase = await createClient();
+    await updateTournamentMutation(supabase, tournamentId, parsed.data);
```
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/actions/tournaments.ts]: Fixed in commit 73edb898 — strip undefined-valued keys from `parsed.data` before the empty-payload check, and pass the stripped object to the mutation. `{ name: undefined }` now short-circuits as a no-op. Tests: "treats explicit-undefined keys as empty and skips the mutation" and "forwards only defined fields to the mutation".

File: apps/web/src/components/tournaments/manage/tournament-settings.tsx [hunk 8]
```diff
@@ -84,72 +121,61 @@ interface TournamentSettingsProps {
...
     maxParticipants: tournament.max_participants?.toString() || "",
-    playerCapEnabled: tournament.max_participants !== null,
-    startDate: tournament.start_date || undefined,
-    endDate: tournament.end_date || undefined,
-    // Registration settings
+    // Loose null-check (`!= null`) so an omitted prop is treated as uncapped,
+    // not as "cap enabled with empty value".
+    playerCapEnabled: tournament.max_participants != null,
+    startDate: tournament.start_date ?? null,
+    endDate: tournament.end_date ?? null,
```
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/components/tournaments/manage/tournament-settings.tsx:150]: Fixed in commit 0d5d78c6 — `buildFormDataFromTournament` now treats `max_participants <= 0` as uncapped, matching the DB convention in the `waitlist_auto_promotion` RPC. Added an `it.each` regression test for 0 and -1.

File: apps/web/src/components/tournaments/manage/tournament-settings.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/components/tournaments/manage/tournament-settings.tsx:345]: Stale — already fixed in commit 9fb011b9 (live formData/phaseConfigs are snapshotted into local consts at save start). The original thread is resolved at https://github.com/thatguyinabeanie/trainers.gg/pull/291#discussion_r3173502427.

File: apps/web/src/actions/tournaments.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/actions/tournaments.ts:199]: Stale — already fixed in commit eb6f28d6 (empty-payload guard short-circuits before the mutation). Tightened further in 73edb898 to also strip explicit-undefined keys.

File: apps/web/src/components/tournaments/manage/tournament-settings.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/components/tournaments/manage/tournament-settings.tsx:229]: Stale — already fixed in commit 73edb898 (`parseStrictInt` helper rejects decimals via `Number.isInteger`).

File: apps/web/src/components/tournaments/manage/tournament-settings.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/components/tournaments/manage/tournament-settings.tsx:756]: Stale — already fixed in commit 73edb898 (same `parseStrictInt` helper applied to the late-round onChange).

File: packages/supabase/src/queries/users.ts [hunk 9]
```diff
@@ -23,16 +31,51 @@ export async function getCurrentUser(supabase: TypedClient) {
...
-  const user = userResult.data;
-  const alt = altResult.data;
+  if (userError || !user) return null;
+
+  let alt: Tables<"alts"> | null = null;
+  if (user.main_alt_id != null) {
+    const { data, error } = await supabase
+      .from("alts")
+      .select("*")
+      .eq("id", user.main_alt_id)
```
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply packages/supabase/src/queries/users.ts:52]: Fixed in commit 9610e133 — added `.eq("user_id", authUser.id)` to the main_alt_id lookup. The double-eq filter rejects rows that don't belong to the authenticated user; when that happens the fallback path picks the user's own oldest alt instead. Added a regression test (`falls back to oldest alt when main_alt_id points at another user's alt`) that fakes a different-user alt at the queried id and asserts the imposter is never surfaced.

File: apps/web/src/actions/tournaments.ts [hunk 10]
```diff
@@ -175,27 +177,43 @@ export async function createTournament(data: {
  */
```
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/actions/tournaments.ts:180]: Fixed in commit 68d4d2e4 — rewrote the `updateTournament` JSDoc to describe the actual behavior (validate + strip undefined, short-circuit empty payload, always invalidate tournament + list + community caches via `invalidateTournamentAndCommunityCaches`).

File: apps/web/src/actions/tournaments.ts [hunk 11]
```diff
@@ -453,8 +471,24 @@ export async function deleteTournament(
   try {
     await rejectBots();
     const supabase = await createClient();
```
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/actions/tournaments.ts:478]: Fixed in commit 68d4d2e4 — rewrote the `deleteTournament` JSDoc to describe the post-delete invalidation path (tournaments list + owning community caches), with a note that invalidation runs *after* the mutation succeeds.

File: apps/web/src/actions/tournaments.ts [hunk 12]
```diff
@@ -453,8 +471,24 @@ export async function deleteTournament(
...
+    // Resolve community + bust caches AFTER the delete succeeds — invalidating
+    // first would evict cache tags even when the mutation throws (RLS, FK,
+    // status≠draft), forcing every subsequent read to repopulate stale data.
+    // We capture the community link before the row is gone so the post-delete
+    // invalidation can still reach the community page.
```
[the remark this answers is not in the document]
Comment (thatguyinabeanie) [reply apps/web/src/actions/tournaments.ts:483]: Fixed in commit 68d4d2e4 — edited the PR description: row 2 of the Fixes table now says "caches now invalidate after a successful delete (so a thrown mutation doesn't evict tags)", matching the implementation and the inline rationale comment.
